File: algorithms/2_medium/C/03393_count_paths_with_the_given_xor_value/dp.cpp
typedef ptrdiff_t isize;

constexpr int max_num = 15;
constexpr int mod = 1'000'000'007;

class Solution {
public:
    int countPathsWithXorValue(const vector<vector<int>>& grid, int k) {
        assert(grid.size() > 0 && grid[0].size() > 0 && k >= 0 && k <= max_num);

        const isize rows = std::ssize(grid);
        const isize cols = std::ssize(grid[0]);
        auto buff = vector<int>((2 * cols * (max_num + 1)), 0);
        auto prev = span<int>(&buff[0], cols * (max_num + 1));
        auto curr = span<int>(&buff[cols * (max_num + 1)], cols * (max_num + 1));

        int curr_xor = 0;
        for (int j = 0; j < cols; j += 1) {
            const int v = grid[0][j];
            assert(v >= 0 && v <= max_num);
            curr_xor ^= v;
            prev[j * (max_num + 1) + curr_xor] = 1;
        }

        for (isize i = 1; i < rows; i += 1) {
            for (isize j = 0; j < cols; j += 1) {
                const int v = grid[i][j];
                assert(v >= 0 && v <= max_num);
                for (int prev_val = 0; prev_val <= max_num; prev_val += 1)  {
                    const int curr_idx = j * (max_num + 1) + (prev_val ^ v);
                    const int up_idx = j * (max_num + 1) + prev_val;

                    curr[curr_idx] = prev[up_idx];
                }

                if (j > 0) {
                    for (int prev_val = 0; prev_val <= max_num; prev_val += 1)  {
                        const int curr_idx = j * (max_num + 1) + (prev_val ^ v);
                        const int left_idx = (j - 1) * (max_num + 1) + prev_val;

                        curr[curr_idx] = (curr[curr_idx] + curr[left_idx]) % mod;
                    }
                }
            }

            std::swap(curr, prev);
            for (auto& v : curr) {
                v = 0;
            }
        }

        const isize target_idx = (cols - 1) * (max_num + 1) + k;
        return prev[target_idx];
    }
};

File: algorithms/2_medium/C/03393_count_paths_with_the_given_xor_value/dp.rs
impl Solution {
    const MAX_NUM: i32 = 15;
    const MOD: i32 = 1_000_000_007;

    pub fn count_paths_with_xor_value(grid: Vec<Vec<i32>>, k: i32) -> i32 {
        assert!(grid.len() > 0 && grid[0].len() > 0 && k >= 0 && k <= Self::MAX_NUM);

        let (rows, cols) = (grid.len(), grid[0].len());

        let mut buff = vec![0_i32; 2 * cols * (Self::MAX_NUM as usize + 1)];
        let (mut prev, mut curr) = buff.split_at_mut(cols * (Self::MAX_NUM as usize + 1));

        let mut xor = 0_i32;
        for (j, &v) in grid[0].iter().enumerate() {
            debug_assert!(v >= 0 && v <= Self::MAX_NUM);
            xor ^= v;
            prev[j * (Self::MAX_NUM as usize + 1) + xor as usize] = 1;
        }

        for line in grid.iter().skip(1) {
            for (j, &v) in line.iter().enumerate() {
                debug_assert!(v >= 0 && v <= Self::MAX_NUM);
                for prev_val in 0..=Self::MAX_NUM  {
                    let curr_idx = j * (Self::MAX_NUM as usize + 1) + (prev_val ^ v) as usize;
                    let up_idx = j * (Self::MAX_NUM as usize + 1) + prev_val as usize;

                    curr[curr_idx] = prev[up_idx];
                }

                if j > 0 {
                    for prev_val in 0..=Self::MAX_NUM  {
                        let curr_idx = j * (Self::MAX_NUM as usize + 1) + (prev_val ^ v) as usize;
                        let left_idx = (j - 1) * (Self::MAX_NUM as usize + 1) + prev_val as usize;

                        curr[curr_idx] = (curr[curr_idx] + curr[left_idx]) % Self::MOD;
                    }
                }
            }

            (prev, curr) = (curr, prev);
            curr.iter_mut().for_each(|x| *x = 0);
        }

        let target_idx = (cols - 1) * (Self::MAX_NUM as usize + 1) + k as usize;
        prev[target_idx]
    }
}
